### ISSUE TITLE
tab switching (preserve tab layout, configure defaults & share links)

### ISSUE THREAD

**Issue by oliver-sanders** [cylc/cylc-ui#662]
> [!NOTE]
> **Definition**: "Workspace"
>
> The part of the UI within the Lumino instance. Basically everything that isn't GScan in the "default" view.
>
> A workspace could be for:
> * An installed workflow (only workspace implemented at the moment).
> * A source workflow *(soon)*.
> * The user dashboard *(one day)*.

The aims the GScan tabbing system are:

* To simplify workspace selection/switching.
* To make the user's workflow between configuring > running > reviewing easy.
* To make it easy to work in a single browser tab.

Here are four steps to achieve these goals in order of priority.

1. Allow the current tab layout to be saved (urgent).
   * Add a component to the workflow toolbar which allows the current tab layout to saved as the default.
   * Apply the default layout when switching workflows.
   * NOTE: configuring multiple view layouts is blocked by #1278
1. Save the layout last used by a workspace (important). #1664
   * When the user navigates away from a workflow, save the tab layout.
   * When the user navigates back to the workflow, load the tab layout.
   * When the user navigates to a new run of the same workflow, copy the layout from the last run of the workflow.
1. Save view state (useful). _[In progress](https://github.com/cylc/cylc-ui/pulls?q=is%3Apr+initialoptions)_
   * Some aspects of the view state should be preserved with the layout.
   * E.G. Any task state filters should be preserved.
   * E.G. The "log" view should remember what log file it is looking at.
   * These things need to be saved as part of the layout information so that we can re-open the layout as it was before.
1. Named layouts (nice-to-have).
   * Allow users to save "named" layouts.
   * Allow users to re-configure the workspace to a previously saved layout.

**Comment by oliver-sanders** [comment]
Long run we should provide:

* A way to "save" the current layout.
* A mechanism for dumping the view's data into the layout (e.g. if the view is filtered by task state, that should be included in the layout)
* A way to "load" a pre-saved layout.
* A configuration for setting the default layout.
* Possibly an additional configuration on each layout to allow setting it as the default layout based on the workflow hierarchy?

E.G. A layout menu icon which provides a "save current layout" button and lists all pre-saved layouts for quick switching.

**Comment by MetRonnie** [comment]
IMO we should split the "Serialise lumino state into local storage" part off into a 1.4.0 issue?

**Comment by oliver-sanders** [comment]
For view state (i.e. options like filtering set by the user in the view), I suggest we collect these options into a single objet in the component `data () {}`. E.G. `this.options.taskFilters`.

Then the serialisation bit can be done at a higher level, extracting them something like `this.$refs[viewID].options`.

View defaults (e.g. you might want new table views to open with some sort order pre-configured) could be provided to the view as a prop e.g. `this.defaultOptions`. We can use the #471 to make these editable in the UserProfile.

**Comment by oliver-sanders** [comment]
(assigning @AaronDCole as #1128 leads in this direction)

**Comment by oliver-sanders** [comment]
See also #1090

**Comment by oliver-sanders** [comment]
FYI, for part (1), Lumino has a `saveLayout` interface.

https://github.com/jupyterlab/lumino/blob/d51b0575c7797769af5a6aabed39500bd494a918/packages/widgets/src/docklayout.ts#L265

**Comment by MetRonnie** [comment]
I've tried that out and it works just like what we want. Will need to figure out how to restore the Tree/Table/etc. views to the restored tabs, which will be empty otherwise.

![demo13](https://github.com/cylc/cylc-ui/assets/61982285/c9491520-a0f6-40be-86fa-0f131f0cd314)

**Comment by oliver-sanders** [comment]
Moved the remainder of point (1) to https://github.com/cylc/cylc-ui/issues/1834

**Comment by oliver-sanders** [comment]
Moved point (4) to https://github.com/cylc/cylc-ui/issues/1835

**Comment by oliver-sanders** [comment]
Added #1837 to capture preserving layouts between browser sessions.
